clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/NettyP2PBlobTransferManager.java

```diff
@@ -39,31 +48,124 @@ public void start() throws Exception {
+   *    After MAX_RETRIES_FOR_BLOB_TRANSFER_PER_HOST times, if still failed to connect, move to the next possible host.
+   * 3. If the connected host does not have the requested file,
+   *    a VeniceBlobTransferFileNotFoundException is thrown, and the process moves on to the next available host.
+   * 4. [Fatal Case] If any unexpected exception occurs, such as InterruptedException, ExecutionException, or TimeoutException
+   *    during the file/metadata transfer, a VeniceException is thrown, and blob transfer is skipped for bootstrapping to save time.
+   * 5. [Fatal Case] If all peers fail to connect or have no snapshot, a VenicePeersNotFoundException is thrown,
+   *    and Kafka is used for bootstrapping instead.
+   *
+   *  success case:
+   *  1. If the blob is successfully fetched from a peer, an InputStream of the blob is returned.
+   *
+   * @param storeName the name of the store
+   * @param version the version of the store
+   * @param partition the partition of the store
+   * @return the InputStream of the blob
+   * @throws VenicePeersNotFoundException
+   */
   @Override
   public CompletionStage<InputStream> get(String storeName, int version, int partition)
       throws VenicePeersNotFoundException {
-    CompletionStage<InputStream> inputStream;
+    // error case 1: no peers are found for the requested blob
     BlobPeersDiscoveryResponse response = peerFinder.discoverBlobPeers(storeName, version, partition);
     if (response == null || response.isError()) {
       throw new VenicePeersNotFoundException("Failed to obtain the peers for the requested blob");
     }
+
     List<String> discoverPeers = response.getDiscoveryResult();
     if (discoverPeers == null || discoverPeers.isEmpty()) {
       throw new VenicePeersNotFoundException("No peers found for the requested blob");
     }
+    LOGGER
+        .info("Discovered peers {} for store {} version {} partition {}", discoverPeers, storeName, version, partition);
+
+    Instant startTime = Instant.now();
     for (String peer: discoverPeers) {
-      try {
-        // TODO: add some retry logic or strategy to choose the peers differently in case of failure
-        // instanceName comes as a format of <hostName>_<applicationPort>
-        String chosenHost = peer.split("_")[0];
-        LOGGER.info("Chosen host: {}", chosenHost);
-        inputStream = nettyClient.get(chosenHost, storeName, version, partition);
-        return inputStream;
-      } catch (Exception e) {
-        LOGGER.warn("Failed to connect to peer: {}", peer, e);
+      int retryCount = 0;
+      while (retryCount < MAX_RETRIES_FOR_BLOB_TRANSFER_PER_HOST) {
+        try {
+          // instanceName comes as a format of <hostName>_<applicationPort>
+          String chosenHost = peer.split("_")[0];
+          LOGGER.info("Attempt {} to connect to host: {}", retryCount + 1, chosenHost);
+          CompletionStage<InputStream> inputStreamStage = nettyClient.get(chosenHost, storeName, version, partition);
```

Comment:
this can be done in on line. I'm also thinking whether we should change the interface return type. Codes are getting verbose unnecessarily. If needed, we can use `CompletionStage` in the future..